### DISCUSSION TITLE
Simple way to do an accurate FaceSwap

### DISCUSSION THREAD

**Discussion by infinity0** [lllyasviel/Fooocus#2644] [Show and tell]
There's a few threads complaining about the FaceSwap quality, because the "obvious" way to do it without going into debugging settings in fact does give bad results that don't resemble the target. There's a few other threads here that give some good tips but they use hard-to-remember debugging options and/or link to a long video.

Instead, **my method here is simple, reliable, and easy-to-remember**:

1. Check both "Input Image" and Advanced at the bottom. Go into Advanced -> Control -> check "Mixing Image Prompt and Vary/Upscale".
2. Go into "Upscale or Variation" and put your input image with the face to be replaced here. Select **Upscale 1.5x or 2x**.
   - Upscale (Fast 2x) seems to completely ignore the face image.
   - Some other guides advise "Vary (Subtle)" but this gives worse results, bad examples further below.
   - Some other guides advise "Inpainting" but this gives worse results, bad examples further below.
3. Go into "Image Prompt", check Advanced inside here, select **FaceSwap**, and put your face image here. Select **Stop At 1.0 and Weight 0.75**. The default value Stop At for 0.9 gives awful results if you actually want your output to look like the target, bad examples further below.
4. Generate!
5. You can also experiment with setting Weight between 0.6 and 1.0 if e.g. your face picture is very bad quality, or if you want its specific features to stand out. Keep **Stop At = 1.0**.

Note: this method (and many other methods) work best as a single Generate step that doesn't mix in other things you might want to be doing. For example, if you're generating a new scene, get that right with the wrong face first, then do this separately.

Examples:

<details><summary>Input image, with face to be replaced, generated by Fooocus</summary>

![2024-03-26_23-04-58_5362](https://github.com/lllyasviel/Fooocus/assets/78398/72fc925a-5601-4080-934f-e05fb7c4b0f9)
</details>

<details><summary>Face image, of Elon Musk found online. I claim fair use here.</summary>

![MS_Musk_Elon_CloseUp](https://github.com/lllyasviel/Fooocus/assets/78398/6169f131-202f-4c7f-8920-38467c568752)
</details>

<details><summary>Output with weight = 0.75</summary>

![2024-03-27_00-44-28_5485](https://github.com/lllyasviel/Fooocus/assets/78398/8fe8b40a-53c8-4e0f-bcdc-66cca721b8b7)
</details>

<details><summary>Output with weight = 1.0 - more aspects of the face image stand out, arguably 0.75 blends in better here. But your use-case may suit weight = 1.0 better.</summary>

![2024-03-27_00-37-46_6184](https://github.com/lllyasviel/Fooocus/assets/78398/17cf97f4-8bcb-4794-a986-d1f727143ee5)
</details>

<details><summary>Output with weight = 0.6 - blends in better than 0.75 but features of the original face also show through. Again, your use-case may suit this better.</summary>

![2024-03-27_13-23-08_3315](https://github.com/lllyasviel/Fooocus/assets/78398/0e8a3ed6-fc59-496c-bd8c-472f8b6a2a01)
</details>

<details><summary>Another output with weight = 0.6 - simply retrying a few times can give you results closer to what you want</summary>

![2024-03-27_13-24-03_9110](https://github.com/lllyasviel/Fooocus/assets/78398/c6e1e480-3eef-48b4-8776-7fb65e5f935a)
</details>

Bad examples:

<details><summary>Output with Stop At = 0.9 - clearly this is not Elon Musk</summary>

![2024-03-27_00-36-30_5749](https://github.com/lllyasviel/Fooocus/assets/78398/fee791fd-0afe-4509-879d-1303ad1d4dcd)
</details>

<details><summary>Output with Vary (Subtle), Stop At = 1.0, Weight = 0.75 - this resembles Elon Musk but the eyes and certain other things are all weird. The results are even worse if your face image is bad quality.</summary>

![2024-03-27_00-46-54_9682](https://github.com/lllyasviel/Fooocus/assets/78398/3a585319-2ae6-4b5a-a3fc-e18bf0d928ea)
</details>

<details><summary>Output with Inpainting (Method: Improve Detail), Stop At = 1.0, Weight = 0.75 - this vaguely resembles Elon Musk but is clearly inferior to Upscale.</summary>

![2024-03-27_01-06-41_7298](https://github.com/lllyasviel/Fooocus/assets/78398/a62da708-63b3-49e1-9730-253f18a526a9)
</details>

**Comment by xhoxye** [comment]
This method may not work for close-up or high-resolution  images
![QQ截图20240327160848](https://github.com/lllyasviel/Fooocus/assets/129571231/be100dbd-5eb7-4331-bb08-73527d3beebd)
![QQ截图20240327160837](https://github.com/lllyasviel/Fooocus/assets/129571231/6ea017c0-a1cf-4bc5-a354-ae643d99be91)
![QQ截图20240327161659](https://github.com/lllyasviel/Fooocus/assets/129571231/70078959-68fa-449e-b89e-ded5987204cb)
![QQ截图20240327162233](https://github.com/lllyasviel/Fooocus/assets/129571231/4521038e-3542-4f5c-acde-0a7b8e98c672)

**Reply by infinity0** [reply]
> This method may not work for close-up or high-resolution images

Most other things I've seen don't work that well for very high-resolution images, because the AI wasn't trained to work with those. You can shrink your images down, do the face swap, then further re-upscale (without face swap) afterwards.

The example outputs you gave really aren't bad though. You could try adjusting the weight a bit higher too. Do you have an alternative method that works better?

**Reply by xhoxye** [reply]
Your method is good, this is the limitation of Fooocus' current algorithm

**Comment by KevinMeinon** [comment]
Do you set it to FaceSwap or ImagePrompt?

**Reply by infinity0** [reply]
Sorry, FaceSwap inside the ImagePrompt tab yes. I'll fix up the main post to be more explicit about this.

**Comment by xhoxye** [comment]
According to the size of the picture is 896x1152, and the standard human body ratio of 7.5 heads, the proportion of the head in the image is only about 150-200 pixels of square.Twice is about 300-400.
200x200 , 400x400
So, make sure that the size of the face being replaced is under 200 pixels, and that you have a higher face size

**Comment by platform-kit** [comment]
Literally none of these look like the intended person. @lllyasviel what would it take to get official support for InstantID? No face swap model out there beats it, and there's already a fork that has implemented it. Would you take a PR?

**Reply by infinity0** [reply]
> Literally none of these look like the intended person

You need to click the triangles to expand the images in the top post, not the images from @xhoxye's later reply.

> there's already a fork that has implemented it.

Link for the rest of us? Seems there is an outstanding issue. https://github.com/lllyasviel/Fooocus/issues/2025

**Comment by KittyDreadful** [comment]
> 1. Go into Advanced -> Control -> check "Mixing Image Prompt and Vary/Upscale".

Disclaimer: new user

I can't find this setting in the Advanced sidebar.  Is this the right place?  

Thanks.

**Reply by ILikeToasters** [reply]
<img width="380" alt="Screenshot 2024-05-10 230610" src="https://github.com/lllyasviel/Fooocus/assets/4357657/09f9a7ef-f7e6-4f5f-a3c6-659805c03944">

**Reply by ILikeToasters** [reply]
Need to check off Developer Debug Mode to find it.

**Comment by Alextest-art** [comment]
I tried the method - it doesn't work. Generates a completely different person. I played with the resolutions and sizes of source photos and target photos, turned off styles, changed/switched everything that can be changed and switched)))
Clarification - this is not suitable for the people I see every day. If you experiment on strangers (or popular people), then this method sometimes works.
Thus, to date, having tried the methods offered on the Internet, I still do not see any simple options for face replacement, except at Lora's training.

**Comment by platform-kit** [comment]
@Alextest-art InstantID works. Fooocus should implement it.

**Comment by parth9361** [comment]
@infinity0 in your method the upscaling 1.5x with face swap and mixing vary subtle setting are you still gonna be able to get high quality results cause i find in backend settings that it only completes the faceswap in just 18 steps

**Comment by JennaSmi** [comment]
Nice one, I will go have a try on the [face swap](https://www.deepswap.ai/) and get back if there's something confused.

**Reply by belst-n** [reply]
Clearly spam...

**Comment by nazrulahmed0888-ctrl** [comment]
UltraHD 4K full-body portrait of a young man walking sideways through a narrow urban alley at night, lit dramatically by firelight from a burning oil barrel to his left. His white shirt is completely soaked and semi-transparent, clinging to his muscular torso. Water drips from his hair and fingertips. The orange firelight casts moving shadows across his face and body. Reflections shimmer on the wet concrete.

**Comment by nazrulahmed0888-ctrl** [comment]
![FB_IMG_1753205392012](https://github.com/user-attachments/assets/7139bd95-5d6c-4c8e-8f2d-bde8443a766b)
UltraHD 4K full-body portrait of a young man walking sideways through a narrow urban alley at night, lit dramatically by firelight from a burning oil barrel to his left. His white shirt is completely soaked and semi-transparent, clinging to his muscular torso. Water drips from his hair and fingertips. The orange firelight casts moving shadows across his face and body. Reflections shimmer on the wet concrete.
